pipeline/aggregate/blocks/ConversationStats.ts

```diff
@@ -80,9 +79,7 @@ const fn: BlockFn<ConversationStats> = (database, filters, common) => {
                 for (let j = i + 1; j < M; j++) {
                     const a = participants[i];
                     const b = participants[j];
-                    if (a === undefined) {
-                        debugger;
-                    }
+                    if (a === undefined) debugger;
```

Comment:
whoops this definitely should not be here

Reply:
I was curious what that was supposed to be for haha. There's a few other places where there seems to be random assertions or errors thrown where I don't see how they could occur, but I didn't want to touch them as I assumed they did something (and I'm guessing I'm right for the other ones)

Reply:
The point of assertions is that they should never happen, otherwise they are exceptions

Reply:
Yes, but usually it's for something that is theoretically possible depending on input, but the program isn't designed with those exceptions in mind. For example, if support was added for multiple Discord guilds (as in, guilds were stored in the database), guild data would be coming from multiple sources which might specify different names for a Discord guild with the same ID. Of course, it's possible that chatlogs could be uploaded which name the guild differently (such as if one is outdated), but if you don't want ChatAnalytics to support that edge-case, you would assert that the guild names were the same. However, I noticed some assertions where I wouldn't have thought they were even theoretically possible.

Reply:
For example, should this not just be `console.assert(this.active, "No active task);`?
https://github.com/mlomb/chat-analytics/blob/main/pipeline/Progress.ts#L78-L81

Reply:
In fact, you use `this.active` instead of `false` here, but you still wrap it in an if-block for the same expression
https://github.com/mlomb/chat-analytics/blob/main/pipeline/Progress.ts#L93-L96

Reply:
And here, the type of `buffer` is already specified to be `ArrayBuffer` via TypeScript, so as far as I understand this assertion is unreachable because an error would be thrown trying to use the constructor in he first place
https://github.com/mlomb/chat-analytics/blob/main/pipeline/serialization/BitStream.ts#L31

Reply:
For me, assert => theorically impossible, exception => theorically possible.

> but if you don't want ChatAnalytics to support that edge-case, you would assert that the guild names were the same.

I dont get it, it wouldn't be an assert or an error/exception, that can happen in that case

> For example, should this not just be `console.assert(this.active, "No active task);`?

Nope, because TypeScript can't know that `this.active` will not be undefined after the assert. Someone could tamper with the `console.assert ` function (for example disable them in production). You must use an if statement, so TypeScript can catch up.

> In fact, you use this.active instead of false here

Yes, that is a typo, it should be `console.assert(this.active, ...)`, doesnt matter that much tbh.

> And here, the type of buffer is already specified to be ArrayBuffer via TypeScript

That assert is unnecessary at this point. At the beginning
https://github.com/mlomb/chat-analytics/blob/8f64224f133b0a2fee79161b5df966a9882e23aa/pipeline/Types.ts#L57 was `any` and if I recall correctly I forgot to add `.buffer` here

https://github.com/mlomb/chat-analytics/blob/8f64224f133b0a2fee79161b5df966a9882e23aa/pipeline/aggregate/Helpers.ts#L12

which cost many hours of debugging.

Reply:
> For me, assert => theorically impossible, exception => theorically possible.

But don't assertions _cause_ exceptions if the expression is false?
```
assert(expression, message)
```js
is pretty much the same as
```js
if (!expression) {
    throw new Error (message);
}
```
but it's a shorter way of writing it

> I dont get it, it wouldn't be an assert or an error/exception, that can happen in that case

In the same way that it's theoretically possible for the JSON file to have to incorrect syntax, but you obviously don't want the program to support JSON files with an incorrect syntax so an error is thrown if that is the case.

Reply:
The top answer to [this StackOverflow question](https://stackoverflow.com/questions/15313418/what-is-assert-in-javascript) is how I view it
> Usually assertions (as they're called) are used only in "testing" or "debug" builds and stripped out of production code.

Reply:
> But don't assertions cause exceptions if the expression is false

What if you do `console.assert = () => {}`?
And `console.assert` only throws an exception in the browser and not in Node.JS

> In the same way that it's theoretically possible for the JSON file to have to incorrect syntax, but you obviously don't want the program to support JSON files with an incorrect syntax so an error is thrown if that is the case.

Yes, thats what exceptions are for. Assertions are for things that can't never ever happen.